lib/_http_client.js

```diff
@@ -504,19 +504,17 @@ ClientRequest.prototype._deferToConnect = function(method, arguments_, cb) {
   // in the future (when a socket gets assigned out of the pool and is
   // eventually writable).
   var self = this;
```

Comment:
I converted the code as you suggested (good catch btw), and that caused the error
```javascript
cannot read property 'writable' of null

if (socket.writable) { // socket used to be self.socket
  ...
```

This seems to stem from the initial `_deferToConnect` call, where [`this.socket` is `undefined`][0]. When the ClientRequest catches the 'socket' event, the `socket` variable in `onSocket()` refers to the initial `undefined` value which was scoped in. As a result, everything blows up when trying to read the `writable` property of that.

A work around that I found was to have `onSocket()` accept a `sock` argument (given by the `socket` event), and update `socket` with that parameter. So this piece of code works:

```javascript
ClientRequest.prototype._deferToConnect = function(method, arguments_, cb) {
  // ...
  var socket = this.socket;
  function callSocketMethod() {
    if (method)
      socket[method].apply(socket, arguments_);
    if (typeof cb === 'function')
      cb();
  }
  var onSocket = function(sock) { // Now accepts the socket from the 'socket' event.
    socket = sock || socket; // Update socket reference.
    if (socket.writable) {
      callSocketMethod();
    } else {
      socket.once('connect', callSocketMethod);
    }
  };
  if (!socket) {
    this.once('socket', onSocket);
  } else {
    onSocket();
  }
};

```

However I'm not completely sure on the repercussions of this. Maybe leaving the existing implementation would be for the better? I'm interested in what you think

[0]: https://github.com/ahoym/node/blob/DRY-ClientRequest-_deferToConnect/lib/_http_client.js#L520-L521